orderer/common/multichannel/registrar.go

```diff
@@ -1017,10 +1017,7 @@ func (r *Registrar) removeSystemChannel() error {
 		}
 		if !isChannelMember {
 			logger.Debugf("Not a member of channel %s, removing it", channel)
-			err := r.removeMember(channel, cs)
-			if err != nil {
-				return errors.WithMessagef(err, "failed to remove channel: %s", channel)
-			}
+			r.removeMember(channel, cs)
```

Comment:
This will remove the channels asynch. That process may take longer then what it takes to remove the system channel, below. One must not call `r.init(r.consenters)` below before all these channels are deleted. 
Moreover, currently, the lower level `r.ledgerFactory` will not allow concurrent `Remove()` anyway. So all theses goroutines will be executed serially anyway.

I would change the flow of the system channel removal:
1. loop to stop all channels, and collect in a slice the app channels the need to be removed.
2. remove the system channel first, synchronously
3. loop to remove app channels, synchronously
4. init all remaining channels, etc.

This way if we fail during system channel removal, a restart will complete the removal of the system channel. If  we fail during the removal of subsequent app channels, we may miss removal of some of the app channels, but that only means they will end up as followers after restart.

Reply:
🧹